src/main/scala/io/iohk/ethereum/network/p2p/messages/PV63.scala

```diff
@@ -1,30 +1,284 @@
 package io.iohk.ethereum.network.p2p.messages
 
 import akka.util.ByteString
+import io.iohk.ethereum.mpt.HexPrefix.{decode => hpDecode, encode => hpEncode}
 import io.iohk.ethereum.network.p2p.Message
 import io.iohk.ethereum.rlp.RLPImplicits._
+import io.iohk.ethereum.rlp.{encode => rlpEncode}
 import io.iohk.ethereum.rlp._
+import org.spongycastle.util.encoders.Hex
+
+import scala.util.Try
 
 object PV63 {
 
   object GetNodeData {
     implicit val rlpEndDec = new RLPEncoder[GetNodeData] with RLPDecoder[GetNodeData] {
       override def encode(obj: GetNodeData): RLPEncodeable = {
         import obj._
-        RLPList(hashes.map(e => RLPValue(e.toArray[Byte])): _*)
+        commonMptHashes: RLPList
       }
 
       override def decode(rlp: RLPEncodeable): GetNodeData = rlp match {
-        case rlpList: RLPList => GetNodeData(rlpList.items.map(e => ByteString(e: Array[Byte])))
+        case rlpList: RLPList => GetNodeData(rlpList.items.map(byteStringEncDec.decode))
         case _ => throw new RuntimeException("Cannot decode GetNodeData")
       }
     }
 
     val code: Int = Message.SubProtocolOffset + 0x0d
   }
 
-  case class GetNodeData(hashes: Seq[ByteString]) extends Message {
+  case class GetNodeData(commonMptHashes: Seq[ByteString]) extends Message {
     override def code: Int = GetNodeData.code
+
+    override def toString: String = {
+      s"""GetNodeData{
+         |hashes: ${commonMptHashes.map(e => Hex.toHexString(e.toArray[Byte]))}
+         |}
+       """.stripMargin
+    }
+  }
+
+  object NodeData {
+    implicit val rlpEndDec = new RLPEncoder[NodeData] with RLPDecoder[NodeData] {
+      override def encode(obj: NodeData): RLPEncodeable = {
+        import obj._
+
+        RLPList(values.map {
+          case Left(node) => RLPValue(rlpEncode(MptNode.rlpEndDec.encode(node)))
+          case Right(byteValue) => RLPValue(byteValue.toArray[Byte])
+        }: _*)
+      }
+
+      override def decode(rlp: RLPEncodeable): NodeData = rlp match {
+        case rlpList: RLPList =>
+          NodeData(rlpList.items.map { e =>
+            Try {
+              val v = rawDecode(e: Array[Byte])
+              Left(MptNode.rlpEndDec.decode(v))
+            }.getOrElse(Right(ByteString(e: Array[Byte])))
+          })
+        case _ => throw new RuntimeException("Cannot decode NodeData")
+      }
+    }
+
+    val code: Int = Message.SubProtocolOffset + 0x0e
+  }
+
+  case class NodeData(values: Seq[Either[MptNode, ByteString]]) extends Message {
+    override def code: Int = NodeData.code
+
+    override def toString: String = {
+      val v = values.map(v => v.fold(_.toString, b => s"BytString(${Hex.toHexString(b.toArray[Byte])})"))
+
+      s"""NodeData{
+         |values: $v
+         |}
+       """.stripMargin
+    }
+  }
+
+  object MptNode {
+    implicit val rlpEndDec = new RLPEncoder[MptNode] with RLPDecoder[MptNode] {
```

Comment:
I would like to keep it separate because it gives us freedom to change our internal implementation of MTP.
In my opinion, despite similarities those are 2 different things
one is content of received message and other is part of MPT that we store